kombu/transport/consul.py

```diff
@@ -0,0 +1,272 @@
+                              acquire=session_id,
+                              value=self.lock_name) is False:
+            LOGGER.info('Could not obtain a lock on key %s', lock_key)
+            return False
+
+        self.queues[queue]['session_id'] = session_id
+
+        return True
+
+    def _release_lock(self, queue):
+        """
+        Try to release a lock. It does so by simply removing the lock key in
+        Consul.
+
+        :param queue: The name of the queue we want to release the lock from
+        :return: None
+        """
+        LOGGER.debug('Removing lock key %s', self._lock_key(queue))
+        self.client.kv.delete(key=self._lock_key(queue))
+
+    def _destroy_session(self, queue):
+        """
+        Destroy a previously created Consul session and release all locks
+        it still might hold
+        :param queue: The name of the Queue
+        :return: None
+        """
+        LOGGER.debug('Destroying session %s', self.queues[queue]['session_id'])
+        self.client.session.destroy(self.queues[queue]['session_id'])
+
+    def _new_queue(self, queue, **_):
+        self.queues[queue] = {'session_id': None}
+        return self.client.kv.put(key=self._key_prefix(queue), value=None)
+
+    def _delete(self, queue, *args, **_):
+        self._destroy_session(queue)
+        del self.queues[queue]
+        self._purge(queue)
+
+    def _put(self, queue, payload, **_):
+        """
+        Put `message` onto `queue`.
+
+        This simply writes a key to the K/V store of Consul
+        """
+        key = '%s/msg/%d_%s' % (self._key_prefix(queue),
+                                int(round(monotonic() * 1000)),
+                                uuid.uuid4())
+
+        if self.client.kv.put(key=key, value=dumps(payload), cas=0) is False:
+            raise ChannelError('Cannot add key {0!r} to consul'.format(key))
+
+    def _get(self, queue, timeout=None):
+        """
+        Get the first available message from the queue
+
+        Before it does so it obtains a lock on the Key/Value store so
+        only one node reads at the same time. This is for read consistency
+        """
+        if self._obtain_lock(queue) is False:
+            raise Empty
+
+        key = '%s/msg/' % self._key_prefix(queue)
+        LOGGER.debug('Fetching key %s with index %s', key, self.index)
+        self.index, data = self.client.kv.get(key=key, recurse=True,
+                                              index=self.index,
+                                              wait=self.timeout)
+
+        try:
+            if data is None:
+                raise Empty
+
+            LOGGER.debug('Removing key %s with modifyindex %s',
+                         data[0]['Key'], data[0]['ModifyIndex'])
+
+            self.client.kv.delete(key=data[0]['Key'],
+                                  cas=data[0]['ModifyIndex'])
+
+            return loads(data[0]['Value'])
+        except TypeError:
```

Comment:
Where is the TypeError here expected to come from?  The error is so generic you can easily end up silencing some important error.